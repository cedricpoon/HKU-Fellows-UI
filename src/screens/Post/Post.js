import React, { Component } from 'react';
import { Dimensions } from 'react-native';
import { Container, Content } from 'native-base';
import { withNavigation } from 'react-navigation';
import { connect } from 'react-redux';

import PostHeaderMenu from './PostHeaderMenu/PostHeaderMenu';
import { mapLayoutToState } from 'hkufui/components/helper';
import { Header, PostFooter } from 'hkufui/components';
import styles from './Styles';

export class Post extends Component {
  constructor(props) {
    super(props);
    this.state = { title: '', headerLayout: { y: 0, height: 0 } };
    this._openHeaderMenu = this._openHeaderMenu.bind(this);
    this._renderHeaderMenu = this._renderHeaderMenu.bind(this);
  }

  _openHeaderMenu() {
    if (this._popup)
      this._popup.toggle();
  }

  _renderHeaderMenu() {
    const { headerLayout } = this.state;
    const { width } = Dimensions.get("window");

    return (
      <PostHeaderMenu
        position={{ x: width, y: headerLayout.y }}
        parentHeight={headerLayout.height}
        onRef={ref => this._popup = ref}
      />
    );
  }

  componentDidMount() {
    const { navigation } = this.props;
    /* all props from <PostPreview /> has been passed */
    if (navigation) {
      this.setState({
        ...navigation.state.params
      });
    }
  }

  render() {
    const { title, subTitle, native } = this.state;

    return (
      <Container>
        {this._renderHeaderMenu()}
        <Header
          title={{
            context: title,
            numberOfLines: 4,
            color: native ? styles.nativeTitle.color : styles.moodleTitle.color
          }}
          subtitle={subTitle && {
            context: subTitle,
            numberOfLines: 3
          }}
          backable
          rightIcon='more'
          onRightPress={this._openHeaderMenu}
          onLayout={mapLayoutToState('headerLayout', this)}
        />
        <Content>
        </Content>
        <PostFooter firstPage />
      </Container>
    );
  }
}

Post.propTypes = {

}

const mapStateToProps = () => ({

});

const mapDispatchToProps = () => ({

})

export default withNavigation(connect(
  mapStateToProps,
  mapDispatchToProps
)(Post));
