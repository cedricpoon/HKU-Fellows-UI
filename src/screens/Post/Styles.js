import { StyleSheet } from 'react-native';

import { logo } from 'hkufui/theme/palette';

export default StyleSheet.create({
  moodleTitle: {
    color: logo.dimmed.blue
  },
  nativeTitle: {
    color: logo.dimmed.green
  }
});
