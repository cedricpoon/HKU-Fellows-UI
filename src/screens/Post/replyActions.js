import { FILL_REPLIES, LOAD_REPLIES, REFRESH_REPLIES } from 'hkufui/src/constants/actionTypes';
import { OK, LOADING } from 'hkufui/src/constants/loadStatus';
import { localize } from 'hkufui/locale';

import MOCK_POSTS from 'hkufui/static/mock/posts';

const locale = localize({ language: 'en', country: 'hk' });

export function onClear() {
  return dispatch => {
    dispatch({ type: LOAD_REPLIES });
  }
}

export function onRefresh({ alert }) {
  return dispatch => {
    alert(locale['comment.refreshing']);
    dispatch({ type: REFRESH_REPLIES })
  }
}

export function onLoad({ id, alert }) {
  return async dispatch => {
    setTimeout(() => {
      dispatch({
        type: FILL_REPLIES,
        payload: { replies: MOCK_POSTS }
      });
      if (alert) {
        alert(locale['comment.refreshed']);
      }
    }, 3000);
  };
}

export default (state = {}, action = {}) => {
  switch (action.type) {
    case FILL_REPLIES:
      return {
        ...state,
        replies: action.payload.replies,
        status: OK
      };
    case LOAD_REPLIES:
      return {
        ...state,
        replies: null,
        status: LOADING
      }
    case REFRESH_REPLIES:
      return {
        ...state,
        status: LOADING
      }
    default:
      return state;
  }
};
