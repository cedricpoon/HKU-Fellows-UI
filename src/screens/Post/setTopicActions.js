import { SET_TOPIC } from "hkufui/src/constants/actionTypes";

export const onSetTopic = (topicId) => ({
  type: SET_TOPIC,
  payload: { topicId }
});

const setTopicActions = (state = {}, action = {}) => {
  switch (action.type) {
    case SET_TOPIC:
      return {
        ...state,
        recentTopic: action.payload.topicId
      }
    default:
      return state;
  }
};

export default setTopicActions;
