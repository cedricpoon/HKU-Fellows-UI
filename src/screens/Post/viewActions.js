import {
  FILL_REPLIES,
  LOAD_REPLIES,
  REFRESH_REPLIES,
  LOAD_REPLIES_FAILED
} from 'hkufui/src/constants/actionTypes';
import { Alert } from 'react-native';
import { OK, LOADING, FAIL } from 'hkufui/src/constants/loadStatus';
import { localize } from 'hkufui/locale';
import { link, view, login, vote } from 'hkufui/config/webapi';
import defaultState from 'hkufui/src/store/globalState';

import { onLogin } from '../Login/authenticate';

const locale = localize({ language: 'en', country: 'hk' });

export function onClear() {
  return dispatch => {
    dispatch({ type: LOAD_REPLIES });
  }
}

export function onRefresh({ alert }) {
  return dispatch => {
    alert(locale['replies.refreshing']);
    dispatch({ type: REFRESH_REPLIES })
  }
}

async function fetchPost({ credential, path, isMoodleKey, payload }) {
  const response = await fetch(link(path), {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      username: credential.userId,
      token: credential.token,
      ...(isMoodleKey && { moodleKey: credential.moodleKey }),
      ...payload
    }),
  });
  return await response.json();
}

export function onVote({ topicId, postId, value, alert }) {
  return async (dispatch, getState) => {
    const { credential } = getState();
    // +1 -> vote up / -1 -> vote down
    const votePath = value > 0 ? vote.up({ postId }) : vote.down({ postId })
    try {
      const res = await fetchPost({ credential, path: votePath });
      if (res.status === 200) {
        // Successfully voted
        dispatch({ type: REFRESH_REPLIES });
        dispatch(onLoad({ id: topicId }));
      } else {
        // failure
        switch(res.status) {
          case 409:
            alert(locale['replies.voted']);
            break;
          case 410:
            alert(locale['replies.selfVote']);
            break;
          default:
            alert(`${locale['replies.voteIssue']} ${res.status}`);
            break;
        }
      }
    } catch (error) {
      alert(locale['replies.voteIssue']);
    }
  }
}

export function onNotify({ topicId }) { // eslint-disable-line
  return dispatch => { // eslint-disable-line
    Alert.alert('onNotify');
  }
}

export function onAccept({ topicId, postId, alert }) {
  return async (dispatch, getState) => {
    const { credential } = getState();
    try {
      const res = await fetchPost({ credential, path: view.adopt({ topicId }), payload: { postId } });
      if (res.status === 200) {
        // Successfully voted
        dispatch({ type: REFRESH_REPLIES });
        dispatch(onLoad({ id: topicId }));
      } else {
        // failure
        alert(`${locale['replies.acceptError']} ${res.status}`);
      }
    } catch (error) {
      alert(locale['replies.acceptError']);
    }
  }
}

export function onLoad({ id, alert }) {
  return async (dispatch, getState) => {
    const { credential } = getState();

    try {
      const res = await fetchPost({ credential, path: view.index({ topicId: id }), isMoodleKey: true });
      if (res.status === 200) {
        const { posts, native: _native, owned: _owned, ...restPayload } = res.payload;
        const native = _native === 1;
        const owned = _owned === 1;
        // success
        dispatch({
          type: FILL_REPLIES,
          payload: { replies: posts, topicInfo: { native, owned, ...restPayload } }
        });
        if (alert) {
          alert(locale['replies.refreshed']);
        }
      } else {
        // failure
        switch(res.status) {
          case 408:
            // refresh token
            await dispatch(onLogin({
              credential: { username: credential.userId, passphrase: credential.passphrase },
              path: login.passphrase()
            }));
            dispatch(onLoad({ id, alert }));
            break;
          default:
            dispatch({ type: LOAD_REPLIES_FAILED });
        }
      }
    } catch (error) {
      dispatch({ type: LOAD_REPLIES_FAILED });
    }
  };
}

export default (state = {}, action = {}) => {
  switch (action.type) {
    case FILL_REPLIES:
      return {
        ...state,
        ...action.payload,
        status: OK
      };
    case LOAD_REPLIES:
      return {
        ...state,
        replies: null,
        topicInfo: defaultState.replies.topicInfo,
        status: LOADING
      }
    case REFRESH_REPLIES:
      return {
        ...state,
        status: LOADING
      }
    case LOAD_REPLIES_FAILED:
      return {
        ...state,
        status: FAIL
      }
    default:
      return state;
  }
};
