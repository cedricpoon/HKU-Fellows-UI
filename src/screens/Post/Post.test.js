import React from 'react';
import { shallow } from 'enzyme';
import { Post } from './Post';
import defaultState from 'hkufui/src/store/globalState';

describe('Testing Post Screen', () => {
  const requiredProps = {
    credential: {},
    encryptor: () => { return '' },
    decryptor: () => { return {} },
    topicInfo: defaultState.replies.topicInfo
  };

  it('renders as expected', () => {
    const wrapper = shallow(
      <Post comments={[]} {...requiredProps} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('is loading', () => {
    const wrapper = shallow(
      <Post {...requiredProps} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
