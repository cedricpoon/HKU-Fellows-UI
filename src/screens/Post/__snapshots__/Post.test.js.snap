// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Post Screen renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Post
    comments={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <PostHeaderMenu
          onRef={[Function]}
          parentHeight={0}
          position={
            Object {
              "x": 750,
              "y": 0,
            }
          }
        />,
        <Header
          animated={false}
          backable={true}
          onLayout={[Function]}
          onRightPress={[Function]}
          rightIcon="more"
          rightStyle={
            Object {
              "color": "#4E4A47",
            }
          }
          subtitle={null}
          title={
            Object {
              "color": "#2F78B0",
              "context": "",
              "numberOfLines": 4,
            }
          }
        />,
        <PostSwipable
          comments={Array []}
          onChangeTab={[Function]}
          onRef={[Function]}
        />,
        <PostFooter
          firstPage={true}
          lastPage={false}
          onPageChangeThunk={[Function]}
        />,
      ],
      "virtual": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "native": undefined,
          "onRef": [Function],
          "parentHeight": 0,
          "position": Object {
            "x": 750,
            "y": 0,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "animated": false,
          "backable": true,
          "onLayout": [Function],
          "onRightPress": [Function],
          "rightIcon": "more",
          "rightStyle": Object {
            "color": "#4E4A47",
          },
          "subtitle": null,
          "title": Object {
            "color": "#2F78B0",
            "context": "",
            "numberOfLines": 4,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "comments": Array [],
          "onChangeTab": [Function],
          "onRef": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "firstPage": true,
          "lastPage": false,
          "onPageChangeThunk": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <PostHeaderMenu
            onRef={[Function]}
            parentHeight={0}
            position={
              Object {
                "x": 750,
                "y": 0,
              }
            }
          />,
          <Header
            animated={false}
            backable={true}
            onLayout={[Function]}
            onRightPress={[Function]}
            rightIcon="more"
            rightStyle={
              Object {
                "color": "#4E4A47",
              }
            }
            subtitle={null}
            title={
              Object {
                "color": "#2F78B0",
                "context": "",
                "numberOfLines": 4,
              }
            }
          />,
          <PostSwipable
            comments={Array []}
            onChangeTab={[Function]}
            onRef={[Function]}
          />,
          <PostFooter
            firstPage={true}
            lastPage={false}
            onPageChangeThunk={[Function]}
          />,
        ],
        "virtual": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "native": undefined,
            "onRef": [Function],
            "parentHeight": 0,
            "position": Object {
              "x": 750,
              "y": 0,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "animated": false,
            "backable": true,
            "onLayout": [Function],
            "onRightPress": [Function],
            "rightIcon": "more",
            "rightStyle": Object {
              "color": "#4E4A47",
            },
            "subtitle": null,
            "title": Object {
              "color": "#2F78B0",
              "context": "",
              "numberOfLines": 4,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "comments": Array [],
            "onChangeTab": [Function],
            "onRef": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "firstPage": true,
            "lastPage": false,
            "onPageChangeThunk": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
